File: lecture05/02_lotto/lotto.cpp
// lotto.cpp

#include <algorithm>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

// Functions

void combinations(vector<int> &s, vector<int> &d, size_t k) {
    // Base: have complete subset
    if (k == d.size()) {
    	if (s.size() == 6) {
	    cout << s[0];
	    for (size_t i = 1; i < s.size(); i++) {
		cout << " " << s[i]; 
	    } 
	    cout << endl;
	}
    	return;
    }
    
    // Recurse: with current
    s.push_back(d[k]);
    combinations(s, d, k + 1);
    s.pop_back(); // Reset subset

    // Recurse: skip current
    combinations(s, d, k + 1);
}

int read_numbers(vector<int> &v) {
    // Read line
    string line;
    if (!getline(cin, line)) {
    	return 0;
    }

    // Read k (first number)
    stringstream ss(line);
    size_t k;
    ss >> k;
    if (!k) {
    	return 0;
    }

    // Read numbers
    v.resize(k);
    for (size_t i = 0; i < k; i++) {
    	ss >> v[i];
    }

    return v.size();
}

// Main execution

int main(int argc, char *argv[]) {
    vector<int> numbers;

    for (size_t test_case = 0; read_numbers(numbers); test_case++) {
    	// TODO: Sort numbers
	sort(numbers.begin(), numbers.end());

	// TODO: Handle blank space between test cases
	if (test_case) {
	    cout << endl;
	}

	// TODO: Generate combinations recursively
	vector<int> combination;
	combinations(combination, numbers, 0);
    }

    return 0;
}

File: lecture05/02_lotto/lotto.py
#!/usr/bin/env python3

import itertools
import sys

# Constants

LOTTO_NUMBERS = 6

# Functions

def display_combinations(s, n=LOTTO_NUMBERS):
    # Use itertools to display combinations
    for combination in sorted(itertools.combinations(s, n)):
        print(' '.join(map(str, combination)))

def combinations(s, d, k):
    # Base: have a complete subset
    if len(d) == k:
        if len(s) == LOTTO_NUMBERS:
            print(' '.join(map(str, s)))
        return

    # Recurse: with current
    combinations(s + [d[k]], d, k + 1)

    # Recurse: skip current
    combinations(s, d, k + 1)

def main():
    for line, numbers in enumerate(map(str.split, sys.stdin)):
        if len(numbers) <= 1:
            break

        if line:
            print()

        # display_combinations(map(int, numbers[1:]))
        numbers = sorted(map(int, numbers[1:]))
        combinations([], numbers, 0)

# Main execution

if __name__ == '__main__':
    main()
